### PR TITLE
Task 3

### PR DESCRIPTION
**Pull request by lerastromtsova**
### Solutions for Chapter 3

Hi @vrom911 @chshersh, this chapter isn't ready yet but if you could take a look at the WIP, it would be great. I think there's something I don't fully understand about types in Haskell. This chapter took me so long, and I'm not 100% sure I made all the tasks right. For some tasks I even have no clue how to properly implement them 😔

So I will be grateful for your comments, hopefully that'll help me do the last task and the Challenge.

### CODE REVIEW

File: src/Chapter3.hs [hunk 1]
```diff
@@ -52,6 +52,7 @@ provide more top-level type signatures, especially when learning Haskell.
 {-# LANGUAGE InstanceSigs #-}
 
 module Chapter3 where
+import Data.Typeable
```
Comment (vrom911) [inline src/Chapter3.hs:55]: Intriguing start!

File: src/Chapter3.hs [hunk 2]
```diff
@@ -343,6 +344,31 @@ Define the Book product data type. You can take inspiration from our description
 of a book, but you are not limited only by the book properties we described.
 Create your own book type of your dreams!
 -}
+data Date = Date
+    { day   :: Int  -- is there a way to restrict acceptable values?
```
Comment (vrom911) [inline src/Chapter3.hs]: That is a very good question!

There are several ways to do that actually.

1. __Smart constructors__. You provide only one way to create the value, through the special function, in which you can make a few validations and checks. You can read a bit more on the pattern here: https://kowainik.github.io/posts/haskell-mini-patterns#smart-constructor 
2. __Custom types__. You can use a custom type to represent that, for example, you can create the sum type for months 🙂 
3. There are very fancy type-level magic tricks, which we will not cover at this point of the course 😄 

If you are interested to know how the types look like in Haskell library, you can check it out here: You can read a bit more on the pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#smart-constructor

File: src/Chapter3.hs [hunk 3]
```diff
@@ -343,6 +344,31 @@ Define the Book product data type. You can take inspiration from our description
...
+    , month :: Int
+    , year  :: Int
+    }
+
+data Author = Author
+    { firstName  :: String
+    , lastName   :: String
+    , middleName :: String -- is there a way to make this optional?
```
Comment (vrom911) [inline src/Chapter3.hs]: Yes, you can use the `Maybe` data type here. We speak a bit about that type later in the chapter, but it exactly means that it either exists or not 🙂 
```suggestion
    , middleName :: Maybe String
```

File: src/Chapter3.hs [hunk 4]
```diff
@@ -343,6 +344,31 @@ Define the Book product data type. You can take inspiration from our description
...
+    { genreName :: String
+    }
+
+data Book = Book
+    { title           :: String
+    , author          :: Author
+    , publicationDate :: Date
+    , genre           :: Genre
+    , language        :: String
+    }
```
Comment (vrom911) [inline src/Chapter3.hs:377]: Very cool type! 👍🏼

File: src/Chapter3.hs [hunk 5]
```diff
@@ -374,6 +400,34 @@ after the fight. The battle has the following possible outcomes:
...
+Monster and Knight types as their properties are the same
+-}
+data Fighter = Fighter
+    { health :: Int
+    , attack :: Int
+    , gold   :: Int
+    } deriving (Show)
+
+{- Returns the updated knight, monster and fight result (-1 knight lost, 1 knight won, 0 draw) -}
+fight :: (Fighter, Fighter)-> (Fighter, Fighter, Int)
```
Comment (vrom911) [inline src/Chapter3.hs:420]: Nice function!
The only issue with the reusage of the same data type for Monster and Knight in here is that you can easily mix them when calling the function:

```haskell
> fight (knight, monster)
> fight (monster, knight)
```

Both of the calls are absolutely valid, but the results are going to be different!

Here, we had in mind to make two type for each Monster and Knight. But as you already noticed, these two would be almost identical, and it feels wrong to duplicate that. Later we show the `newtypes` which usually resolve these issues (we didn't expect you to use them in here, this is just the additional note 🙂 )
```haskell
newtype Monster = Monster Fighter
newtype Knight = Knight Fighter
```

File: src/Chapter3.hs [hunk 6]
```diff
@@ -374,6 +400,34 @@ after the fight. The battle has the following possible outcomes:
...
+fight (knight, monster)
+    | health knight  == 0 && health monster > 0 = (knight, monster, -1)
```
Comment (vrom911) [inline src/Chapter3.hs]: You can check on helath `<= 0` to make it less strict 👍🏼

File: src/Chapter3.hs [hunk 7]
```diff
@@ -374,6 +400,34 @@ after the fight. The battle has the following possible outcomes:
...
+    where
+        nextRound :: (Fighter, Fighter) -> (Fighter, Fighter)
+        nextRound (knight, monster) = (updatedKnight, updatedMonster)
+            where
+                updatedMonster :: Fighter
+                updatedKnight  :: Fighter
+                updatedMonster = monster { health = health monster - attack knight }
+                updatedKnight = if health monster > 0
+                                then knight { health = (health knight - attack monster) }
+                                else knight
```
Comment (vrom911) [inline src/Chapter3.hs]: You can simplify it a bit and leave just `updateMonster` and `uodateKnight` functions:

 ```haskell
    | otherwise = fight (updatedKnight, updatedMonster)
  where
    updatedMonster :: Fighter
    updatedMonster = monster { health = health monster - attack knight }
            
    updatedKnight  :: Fighter
    updatedKnight = 
        if health monster > 0
        then knight { health = (health knight - attack monster) }
        else knight
```

File: src/Chapter3.hs [hunk 8]
```diff
@@ -460,6 +514,24 @@ Create a simple enumeration for the meal types (e.g. breakfast). The one who
...
+    | EnglishBreakfast
+        { tea       :: Int
+        , milk      :: Int
+        , friedEggs :: Int
+        , toasts    :: Int
+        }
+    | RussianBreakfast
+        { caviar    :: Int
+        , pancakes  :: Int
+        }
```
Comment (vrom911) [inline src/Chapter3.hs]: Yummy 😋 🥞 Very creative 🎨 

Nice usage of both sums and record types!

One note in here, record inside the sum type is considered an antipattern in Haskell. This is because it produces partial functions.
As we mentioned, records bring filed names as function in scope.
In case of a record, for example, we would have the function:

```haskell
jam :: Breakfast -> Int
```

However, this function is partial, as it will fail in the runtime on such input:

```haskell
jam (RussianBreakfast 1 2)
```

The solution here is to move the record into a separate data type and use it inside the sum type as unnamed arg:

```haskell
data FrenchBreakfast = FrenchBreakfast
    { croissant :: Int    
    , jam       :: Int
    , butter    :: Int
    }
...

data Breakfast
    = French FrenchBreakfast
    | English EnglishBreakfast
    ...
```

File: src/Chapter3.hs [hunk 9]
```diff
@@ -480,6 +552,80 @@ After defining the city, implement the following functions:
...
+        , houses  :: [House]
+        }
+    | CityWithChurch
+        { church  :: Church
+        , houses  :: [House]
+        }
+    | CityWithLibrary
+        { library :: Library
+        , houses  :: [House]
+        }
```
Comment (vrom911) [inline src/Chapter3.hs]: Interesting approach of using a sum type for the city! 👍🏼 🧠 

I feel like it would be more convenient to work `City` if it was a record type instead of a sum type.

First, we need to decide with the castle. You have walls and castles separate, but that could lead to some problems.this also allows having some states that should not ideally exist: like having walls without any castle at all. With this representation, this is possible. However, we actually can restrict this on type-level by providing more representative types 🙂 This is actually the greatest power of Haskell and type system 😍 

This pattern is called "Make illegal states unrepresentable", and you can read more about it here:

* https://kowainik.github.io/posts/haskell-mini-patterns#make-illegal-states-unrepresentable

 Let's look at the task:

> Wall, but only if the city has a castle
> -- Optional castle with a name (as 'String')

That means that walls are very connected to the castle, and they could not exist separately. So we have three options for the castle:

1. Castle with a name
2. Castle with the name + walls
3. No castle

This looks like a sum type!:

```haskell
data Castle 
    = Castle String
    | WallCastle String
    | None
```

Note that your idea with castle suits here perfectly as well. But you need to somehow connect them together, for example another option could be:

```haskell
data CastleBuilding = Castle Building
    { castleName :: Castle
    , castleWall :: [Wall] 
    }
```

And in the city, we will have `Maybe CastleBuilding`.


Also, you see that only Church or only Library are allowed in the city. This is a good use case for a sum type, as it represents "OR" pattern very well:

```haskell
data ChurchOrLibrary
    = Church
    | Library
```

And you won't need to create additional types separately for each building, these constructors are enought.

So, in summary, instead of having two top-level constructors and duplicating all fields on both of them, you can define `City` as a record with optional castle:

After all that is done, Then let's build the city:

```haskell
data City = City
    -- Optional castle with a __name__ (as 'String')
    -- Wall, but only if the city has a castle
    { cityCastle :: Castle -- from the above
    -- OR you can do it with the other Castle type
    -- cityCastle :: Maybe Castle
    -- Church or library but not both
    , cityChurchOrLibrary :: ChurchOrLibrary
    -- any number of houses. Each house has one, two, three or four __people__ inside.
    , cityHouse s:: [House]
    }
```

File: src/Chapter3.hs [hunk 10]
```diff
@@ -480,6 +552,80 @@ After defining the city, implement the following functions:
...
+
+buildHouse  :: City -> Int -> City
+buildHouse city numPersons = city { houses = (new_house:houses city)}
+    where
+        new_house
+            | numPersons == 1 = House OnePerson
+            | numPersons == 2 = House TwoPersons
+            | numPersons == 3 = House ThreePersons
+            | numPersons == 4 = House FourPersons
+            | otherwise = error "Invalid number of persons"
```
Comment (vrom911) [inline src/Chapter3.hs]: Nice smart constructor in here! 👍🏼 

Note that you could use the `case-of` construction in here:

```haskell
        newHouse = case numPersons of
            1 -> House OnePerson
...
```

It is usually not that good to have `error` in code. It is better to use more typesafe interfaces. 
For example in here you can create the function like this:

```haskell
makeHouse :: Int -> Maybe House
makeHouse numPersons = case numPersons of
    1 -> Just (House OnePerson)
    2 -> ...
    3 -> ...
    4 -> ...
    _ -> Nothing
```

And in `buildHouse` you can pattern match on the result of `makeHouse` and add the house only in case of `Just`. In case of `Nothing you could return unchanged city 🙂

File: src/Chapter3.hs [hunk 11]
```diff
@@ -560,22 +706,31 @@ introducing extra newtypes.
...
 
-calculatePlayerDefense :: Int -> Int -> Int
-calculatePlayerDefense armor dexterity = armor * dexterity
+calculatePlayerDefense :: Armor -> Dexterity -> Defense
+calculatePlayerDefense (Armor armor) (Dexterity dexterity) = Defense (armor * dexterity)
 
-calculatePlayerHit :: Int -> Int -> Int -> Int
-calculatePlayerHit damage defense health = health + defense - damage
+calculatePlayerHit :: Damage -> Defense -> Health -> Health
+calculatePlayerHit (Damage damage) (Defense defense) (Health health) = Health (health + defense - damage)
```
Comment (vrom911) [inline src/Chapter3.hs:731]: Nice work on newtypes! :medal_sports: Newtypes are one very important feature in Haskell :+1:
You can read more about the newtype pattern in our blog post: https://kowainik.github.io/posts/haskell-mini-patterns#newtype

File: src/Chapter3.hs [hunk 12]
```diff
@@ -910,6 +1073,20 @@ Implement instances of "Append" for the following types:
 class Append a where
     append :: a -> a -> a
 
+instance Append [a] where
+    append :: [a] -> [a] -> [a]
+    append x y = x ++ y
```
Comment (vrom911) [inline src/Chapter3.hs]: You can eta-reduce the implementation of `append` for brevity. So you can see how easy it is to create instances :slightly_smiling_face: 
```suggestion
    append = (++)
```

File: src/Chapter3.hs [hunk 13]
```diff
@@ -910,6 +1073,20 @@ Implement instances of "Append" for the following types:
...
+
+newtype Gold = Gold Int
+
+instance Append Gold where
+    append :: Gold -> Gold -> Gold
+    append (Gold x) (Gold y) = Gold (x+y)
```
Comment (vrom911) [inline src/Chapter3.hs]: Great! Note that binary operators should be separated with spaces from the arguments 👾 
```suggestion
    append (Gold x) (Gold y) = Gold (x + y)
```

File: src/Chapter3.hs [hunk 14]
```diff
@@ -910,6 +1073,20 @@ Implement instances of "Append" for the following types:
...
+
+-- didn't quite understand what is required in this task :/
+--instance Append (Maybe a) where
+--    append :: (Maybe a) -> (Maybe a) -> (Maybe a)
```
Comment (vrom911) [inline src/Chapter3.hs]: Yes, here is exactly the function that needs to be written.
Below you absolutely correctly cover the case when both of the arguments are `Just`s. As maybe id a sum type, there are 4 combinations that could potentially be:

```haskell
Just x  <->  Just y
Nothing <-> Just y
Just x  <-> Nothing
Nothing <-> Nothing
```
So the function pattern matching should cover them all.
Let me know if it makes more sense now, or you have a question about the meaning of such instance 🙂

File: src/Chapter3.hs [hunk 15]
```diff
@@ -971,6 +1148,33 @@ implement the following functions:
...
+
+isWeekend :: DayOfWeek -> Bool
+isWeekend x = case x of
+    Saturday -> True
+    Sunday -> True
+    _ -> False
+
+nextDay :: DayOfWeek -> DayOfWeek
+nextDay x = case x of
+    Sunday -> Monday
```
Comment (vrom911) [inline src/Chapter3.hs]: That is right!
You could also use the instance of `Bounded` in here:

```haskell
    if x == maxBound then minBound ...
```

This is also a bit more maintainable. If you decide, for example, to have the "American" system of week days in the data type constructors, you won't need to change this function at all! 🙂

File: src/Chapter3.hs [hunk 16]
```diff
@@ -971,6 +1148,33 @@ implement the following functions:
...
+    _ -> succ x
+
+daysToParty :: DayOfWeek -> Int
+daysToParty x = case x of
+    Saturday -> 6
+    Sunday -> 5
+    _ -> length (enumFromTo x Friday) - 1
```
Comment (vrom911) [inline src/Chapter3.hs:1175]: That is right that you need to use the instance of `Enum` here. Note that is you use `mod` with the number of days in the week you won't need to have any special checks 🙂

**Review by vrom911 — APPROVED** [review]
That a fantastic job you did here. You construct really nice data types 👷🏼 

I understand, that this chapter could be hard, as types in Haskell is the most challenging topic, as it allows so many variations. But you already did great! I wrote a few suggestions, let me know if it makes sense. Also don't hesitate to ask any questions! We are here to help🤗

File: src/Chapter3.hs (hunk 14 above)
Comment (lerastromtsova) [reply src/Chapter3.hs]: Hi @vrom911! Thanks for the feedback 😊 Yes, it does make sense now, I have fixed the function to work with all the cases.

File: src/Chapter3.hs (hunk 15 above)
Comment (lerastromtsova) [reply src/Chapter3.hs]: Sounds reasonable, but when I changed the code to this
```
nextDay :: DayOfWeek -> DayOfWeek
nextDay x = case x of
    maxBound -> minBound
    _ -> succ x
```
The compiler gives a warning "Pattern match is redundant". Although in fact it's not 🤔 If we remove the part `maxBound -> minBound`, then calling the function on Sunday will result in an error.
Is there a way to somehow tell the compiler that the warning should be suppressed?

File: src/Chapter3.hs [hunk 17]
```diff
@@ -603,28 +598,33 @@ countPeople house = case people house of
...
-    | otherwise = error "Not enough population :("
+buildHouse city numPersons = case new_house of
+    Just house -> city { cityHouses = (house:cityHouses city)}
+    Nothing -> city
     where
-        population :: Int
-        population = sum (map countPeople (houses city))
+        new_house = makeHouse numPersons
+
+---- Do not know exactly how to check that a city has a castle
```
Comment (vrom911) [inline src/Chapter3.hs]: You can use `case-of` here:

```haskell
... = case cityCastle city of
    Castle castle -> city { cityCastle = WallCastle castle}
    _ -> city
```

So you actually need to do something only when there is castle without walls, in all other cases, nothing could be done 🙂

File: src/Chapter3.hs [hunk 18]
```diff
@@ -1075,18 +1075,20 @@ class Append a where
...
--- didn't quite understand what is required in this task :/
---instance Append (Maybe a) where
---    append :: (Maybe a) -> (Maybe a) -> (Maybe a)
---    append (Just x) (Just y) = Just (append x y)
+instance Append (Maybe a) where
+    append :: (Maybe a) -> (Maybe a) -> (Maybe a)
+    append (Just x) (Just y) = append (Just x) (Just y)
+    append (Nothing) (Just y) = Just y
+    append (Just x) (Nothing) = Just x
+    append (Nothing) (Nothing) = Nothing
```
Comment (vrom911) [inline src/Chapter3.hs]: This is an absolutely correct solution!

However, we can notice that these 3 cases could be reduced to 2!
Let's see what we can unite in here.

If some of the arguments are `Nothing`, you actually always return the second one! So you need to inspect only two cases: `Nothing` in the first position, and `Nothing` on the second one 🙂

**Review by vrom911 — APPROVED** [review]
> Sounds reasonable, but when I changed the code to this
> 
> nextDay :: DayOfWeek -> DayOfWeek
> nextDay x = case x of
>     maxBound -> minBound
>     _ -> succ x
> The compiler gives a warning "Pattern match is redundant". Although in fact it's not 🤔 If we remove the part maxBound -> minBound, then calling the function on Sunday will result in an error.
> Is there a way to somehow tell the compiler that the warning should be suppressed?

It is actually not possible to pattern match on function. `maxBound` is the function, so compiler can't have it in there.
As Haskell allows names-shadowing, here it assumes, that `maxBound` is just the name that you gave for your WeekDay.
Your function, actually never reaches the second case and will always return `minBound`, as the variable matches any WeekDay 🙂 
Hope this answers the question 👌🏼

File: src/Chapter3.hs (hunk 18 above)
Comment (lerastromtsova) [reply src/Chapter3.hs]: @vrom911, hi! Thank you for the comments. But here, if I remove 
`append (Nothing) (Nothing) = Nothing`
then when I try to call
`append Nothing Nothing`
GHCi says: "Non-exhaustive patterns in function append". Or did you mean something else when you said that these 3 cases could be reduced to 2?

Comment (vrom911) [reply src/Chapter3.hs]: Oh, I see.
Yes, in this case it is not enough to remove the last case, but the previous two should be patched a bit. Specifically, to show that not only `Just` would work this way with `Nothing` but anything:

```haskell
    append Nothing y = y
    append x Nothing = x
```

Then it would be the same as the correct solution that you wrote in the first place 👍🏼
